source/grid/grid_generator.cc

```diff
@@ -4240,6 +4252,102 @@ namespace GridGenerator
+    if (n_shells == 0)
+      return; // empty Triangulation
+
+    std::vector<double> radii;
+    radii.push_back(inner_radius);
+    for (unsigned int shell_n = 1; shell_n < n_shells; ++shell_n)
+      if (skewness == 0.0)
+        // same as below, but works in the limiting case of zero skewness
+        radii.push_back(inner_radius +
+                        (outer_radius - inner_radius) *
+                          (1.0 - (1.0 - double(shell_n) / n_shells)));
+      else
+        radii.push_back(
+          inner_radius +
+          (outer_radius - inner_radius) *
+            (1.0 - std::tanh(skewness * (1.0 - double(shell_n) / n_shells)) /
+                     std::tanh(skewness)));
+    radii.push_back(outer_radius);
+
+    auto min_line_length = [](const Triangulation<dim> &tria) -> double {
+      double length = std::numeric_limits<double>::max();
+      for (const auto cell : tria.active_cell_iterators())
+        for (unsigned int n = 0; n < GeometryInfo<dim>::lines_per_cell; ++n)
+          length = std::min(length, cell->line(n)->diameter());
+      return length;
+    };
+
+    double grid_vertex_tolerance = 0.0;
+    for (unsigned int shell_n = 0; shell_n < radii.size() - 1; ++shell_n)
+      {
+        Triangulation<dim> current_shell;
+        GridGenerator::hyper_shell(current_shell,
+                                   center,
+                                   radii[shell_n],
+                                   radii[shell_n + 1],
+                                   n_cells == 0 ? (dim == 2 ? 8 : 12) :
+                                                  n_cells);
+
+        // The innermost shell has the smallest cells: use that to set the
+        // vertex merging tolerance
+        if (grid_vertex_tolerance == 0.0)
+          grid_vertex_tolerance = 0.5 * min_line_length(current_shell);
+
+        Triangulation<dim> temp(std::move(triangulation));
+        triangulation.clear();
+        GridGenerator::merge_triangulations(current_shell,
+                                            temp,
+                                            triangulation,
+                                            grid_vertex_tolerance);
+      }
+
+    const types::manifold_id manifold_id = 0;
+    triangulation.set_all_manifold_ids(manifold_id);
+    if (dim == 2)
+      triangulation.set_manifold(manifold_id, PolarManifold<dim>(center));
+    else if (dim == 3)
+      triangulation.set_manifold(manifold_id, SphericalManifold<dim>(center));
+
+    if (colorize)
+      for (const auto &cell : triangulation.active_cell_iterators())
+        for (unsigned int face_n = 0;
+             face_n < GeometryInfo<dim>::faces_per_cell;
+             ++face_n)
+          {
+            auto face = cell->face(face_n);
+            if (face->at_boundary())
+              {
+                if ((face->center() - center).norm() <
+                    (inner_radius + outer_radius) / 2.0)
```

Comment:
I am not sure I understand where this could be problematic. Since this is for the coarsest level, we are guaranteed (since the center is not calculated along the geodesic) that `face->center() - center).norm() < inner_radius`. I added the averaging to make this more robust.

> I find that checking if the vertex lies on the radius often works best for these limiting cases.

This is better anyway (its a lot simpler!). Do you have a recommendation for the tolerance? I am thinking of rewriting this as
```cpp
  if (std::abs(face->vertex(0) - inner_radius) < inner_radius*1e-14) 
```

Is this what you have done before?

Reply:
I must admit that I've not tested my implementations for a wide range of conditions, but I think that `if (std::abs(face->vertex(0) - inner_radius) < 1000*std::numeric_limits<double>::epsilon())` would be appropriate, since we're checking that they are approximately the same value. Do you agree?

Reply:
^ I wrote this before checking the changes you'd made. I'd suggest maybe changing your condition to not include the factor `inner_radius`. If its very small (e.g.  `inner_radius=1e-9`) then perhaps this accuracy of `100*eps*inner_radius` is unobtainable.

Reply:
Hm. I want to think some more about how to set up a good error tolerance here.